drivers/base/regmap/regmap-sdw-mbq.c

```diff
@@ -9,35 +10,95 @@
 #include <linux/soundwire/sdw_registers.h>
 #include "internal.h"
 
+struct regmap_mbq_context {
+	struct device *dev;
+
+	struct regmap_sdw_mbq_cfg cfg;
+
+	int val_size;
+};
+
+static int regmap_sdw_mbq_size(struct regmap_mbq_context *ctx, unsigned int reg)
+{
+	int size = ctx->val_size;
+
+	if (ctx->cfg.mbq_size) {
+		size = ctx->cfg.mbq_size(ctx->dev, reg);
+		if (size > ctx->val_size)
+			return -EINVAL;
+	}
+
+	return size;
+}
+
 static int regmap_sdw_mbq_write(void *context, unsigned int reg, unsigned int val)
 {
-	struct device *dev = context;
+	struct regmap_mbq_context *ctx = context;
+	struct device *dev = ctx->dev;
 	struct sdw_slave *slave = dev_to_sdw_dev(dev);
+	int mbq_size = regmap_sdw_mbq_size(ctx, reg);
 	int ret;
 
-	ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg), (val >> 8) & 0xff);
-	if (ret < 0)
-		return ret;
-
-	return sdw_write_no_pm(slave, reg, val & 0xff);
+	switch (mbq_size) {
+	case 4:
+		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+				      (val >> 24) & 0xff);
+		if (ret < 0)
+			return ret;
+
+		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+				      (val >> 16) & 0xff);
+		if (ret < 0)
+			return ret;
+		fallthrough;
+	case 2:
+		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+				      (val >> 8) & 0xff);
+		if (ret < 0)
+			return ret;
+		fallthrough;
+	case 1:
+		return sdw_write_no_pm(slave, reg, val & 0xff);
```

Comment:
Maybe I need more coffee while reading this post-lunch, but now I wonder if we actually need two regmaps for 'regular' 8 bits registers and extended MBQ for more that one byte.

Could we unify the two existing regmap handling?

Reply:
Oops missed this first comment, yeah my intention is that this patch unifies them and allows the MBQ register map to access 8-bit registers.

Reply:
so there would be an additional step of conversion of all existing regmap-sdw.c users.

Or the other way around, maybe the easiest path would be to move all the code from regmap-sdw-mbq.c into regmap-sdw.c, and then transition the small set of users of the MBQ stuff?

Reply:
I don't think we should get rid of the standard sdw regmap stuff, they serve different purposes. The regular map provides access to larger registers without MBQ, that isn't possible through the MBQ map, see all the 32-bit non-soundwire/SDCA registers on the cs42l43 or cs35l56.

Nothing should really "need" converted here, these changes should be entirely backwards compatible. So all existing code will work as is, although I would probably advise that the few codecs using multiple register maps for SDCA do get converted just because of the complexity and error prone-ness of implementing things that way, but that is up to the maintainers of those drivers.

Reply:
not following the point about serving different purposes. I don't see much of a difference between a vendor-specific  and a class-defined register, and I don't like giving people the option of hurting themselves. I don't also understand the "32-bit non-soundwire/SDCA registers on the cs42l43 or cs35l56", the only thing the regular regmap will support is 8-bit registers.

Reply:
There is no hurting ones self here, if you want MBQ features use that regmap if you don't don't. There is no massive problem if you select either option and the cost of switching later is small.

We expanded the regular map to support arbitrary register sizes quite some time ago as part of supporting cs42l43. 522272047dc6 ("regmap: sdw: Remove 8-bit value size restriction")

Merging the two regmaps will make things very complicated and I really see no value to doing so.

Reply:
Loosely the way I see this is the MBQ register map is the SDCA regmap and the plain sdw one is just for plain soundwire devices.

Reply:
that's not my interpretation at all. the plain sdw regmap was for 8 bit registers and larger registers had to use the mbq stuff.

by removing the size restrictions on all sides, Cirrus Logic created an interesting question of what regmap should be used when....

If both implementations of regmaps for soundwire support multiple sizes, I am not sure why we need 2.
The only thing the MBQ does is write the 8-bit values in a specific order to be standard, I can't think how this would impact access to non-SDCA registers?

Reply:
The two register maps access data different for >1 bytes data. The sdw register maps just accesses the bytes sequentially, the mbq register maps uses the mbq protocol whilst accessing them. The mbq map also processes the whole thing as single byte calls to the soundwire core, whereas the regular map passes the whole transaction to the core allowing for slight efficiency improvements that are not possible in the mbq case.

The regular map is more of a direct mapping of the Linux soundwire API to a regmap, whereas the mbq one layers protocol on top. I am not sure I see that it creates any confusion other than arguably in the single byte case without defers where you could legitimately use either map but in that case it just doesn't really matter which you use. Although the standard would seem logical since you are using no mbq features. For higher sized registers only one of the two maps will actually work for a given device. But the differentiation between them seems pretty clear cut to me if you just want raw bytes over soundwire normal map, if you want mbq protocol on top of that mbq.

Reply:
I am probably missing something or not up-to-speed on all this.
Do you have any examples where the regmap stuff updates multiple registers with "slight efficiency improvements that are not possible in the mbq case".
Also if you have an example of non-mbq registers with more than 8 bits, that would be useful.

Reply:
For examples of non-MBQ registers with more than 8-bits you can look at any of cs42l43 registers (they are all 32-bit, non-mbq) the regmap_config is defined in drivers/mfd/cs42l43.c. I would be up for an actual chat if you want to go through the details of cs42l43s hardware more closely, that is probably best done offline. But I guess there are few reasons for this, but mostly Cirrus has been making soundwire chips a lot longer than MBQ has been a thing and our chips tend to offer multiple control buses (i2c/spi).

For the discussions of efficiency. A lot of it is really that the MBQ regmap is forced to use the reg_read/reg_write interface to regmap which processes a single transaction. This forces regmap to do additional processing/copies, particularly for larger bulk/raw read/writes, although those are less likely in an MBQ context anyway.
Secondly, the regmap-sdw-mbq is forced to use sdw_read, also because of the address changing between bytes of a single transaction. But regmap-sdw can use sdw_nwrite to pass the whole tranaction to the soundwire master at once letting it fill its whole internal FIFO and let rip. Using nwrite may also have some benefits if we were to look at integrating BRA into the system for larger transactions, I know you are currently favouring a separate API for BRA and I haven't spent enough time working through it to really form a solid opinion there yet, but it would be nice to not close doors earlier than we need to.

But fundamentally I think those things are fringe benefits and it really comes down to this, the two register maps access multiple byte quantities in different ways, so they should be different register maps. A chip either does things one way or the other way so its pretty easy to pick which map to use, and even if you get it wrong changing between costs a few minutes. Allowing them both to support 8-bit registers, even though those are handled identically in both, also makes sense because it avoid having to write drivers that have multiple regmaps which is error prone and really balloons code size when interfacing with ASoC.